compiler/rustc_middle/src/ty/context.rs

```diff
@@ -1383,32 +1388,94 @@ impl<'tcx> TyCtxt<'tcx> {
         )
     }
 
+    /// Create a new definition within the incr. comp. engine.
+    pub fn create_def(
+        self,
+        parent: LocalDefId,
+        data: hir::definitions::DefPathData,
+        expn_id: rustc_span::ExpnId,
+        span: Span,
+    ) -> LocalDefId {
+        // The following call has the side effect of modifying the tables inside `definitions`.
+        // These very tables are relied on by the incr. comp. engine to decode DepNodes and to
+        // decode the on-disk cache.
+        let def_id = self.definitions.write().create_def(parent, data, expn_id, span);
```

Comment:
There is something *very* subtle going on here: This modifies the `Definitions` table which contains the sources for the `source_span` and `expn_that_defined` queries. So we are basically writing to the query tables of these queries here -- not literally but it amounts to the same thing. 

That should not be a problem because those queries are `eval_always`. But I think it would not even a problem otherwise: it looks like this is essentially the case that Niko described for Salsa 2, i.e. it is OK for a query A to write directly to the tables of other queries if the query key for those other entries contains something that A has created.

I think we should document this somewhere. Either here or on the queries in question.